File: apps/web-jogmania/repo/api/app/services/loot.py
import random

LOOT_TABLE = {
    "common": [
        ("Static Charm", "Keeps the CRT humming."),
        ("Jungle Patch", "Rough fabric from the canopy."),
        ("Pace Badge", "Proof of steady tempo.")
    ],
    "rare": [
        ("Neon Anklet", "Crackles with 80s electricity."),
        ("Vine Runner Patch", "Grants extra airtime."),
        ("Turbo Sweatband", "Feels like a power-up.")
    ],
    "epic": [
        ("Prism Relic", "Splits light into speed boosts."),
        ("CRT Crown", "Arcade royalty status."),
        ("Sunflare Totem", "Burns away fatigue.")
    ]
}


def roll_loot(distance_m: float, duration_s: int, avg_pace_s_per_km: int) -> list[dict]:
    score = distance_m / 200 + max(0, 600 - avg_pace_s_per_km) / 200
    rng = random.Random(f"{distance_m}-{duration_s}-{avg_pace_s_per_km}")

    rolls = 1 + (1 if score > 5 else 0) + (1 if score > 9 else 0)
    items: list[dict] = []

    for _ in range(rolls):
        rarity = "common"
        if score > 8 and rng.random() > 0.6:
            rarity = "epic"
        elif score > 4 and rng.random() > 0.5:
            rarity = "rare"

        name, description = rng.choice(LOOT_TABLE[rarity])
        items.append({"name": name, "rarity": rarity, "description": description})

    return items

File: apps/web-jogmania/repo/repo/apps/web/src/lib/loot.ts
import type { RunSummary } from "@/game/RunnerGame";

export type LootItem = {
  name: string;
  rarity: "common" | "rare" | "epic";
  description: string;
};

const COMMON_LOOT: LootItem[] = [
  { name: "Static Charm", rarity: "common", description: "Keeps the CRT humming." },
  { name: "Glow Band", rarity: "common", description: "A steady pulse of jungle light." },
  { name: "Arcade Token", rarity: "common", description: "Trade for a pace boost later." }
];

const RARE_LOOT: LootItem[] = [
  { name: "Jungle Prism", rarity: "rare", description: "Splits neon light into pace boosts." },
  { name: "Pulse Capsule", rarity: "rare", description: "Stabilizes streaks under pressure." },
  { name: "Relic Compass", rarity: "rare", description: "Points toward hidden shortcuts." }
];

const EPIC_LOOT: LootItem[] = [
  { name: "Turbo Idol", rarity: "epic", description: "Turns every split into a highlight reel." },
  { name: "Laser Crown", rarity: "epic", description: "Earned by the fastest jungle runners." },
  { name: "Void Lantern", rarity: "epic", description: "Lights the course when the pace surges." }
];

function pickFrom(list: LootItem[]) {
  return list[Math.floor(Math.random() * list.length)];
}

export function rollLoot(summary: RunSummary, sessionPoints: number): LootItem[] {
  const paceScore = summary.avg_pace_s_per_km ? Math.max(0, 420 - summary.avg_pace_s_per_km) : 0;
  const distanceScore = summary.distance_m / 100;
  const totalScore = paceScore + distanceScore + sessionPoints * 0.8;

  if (totalScore > 380) {
    return [pickFrom(EPIC_LOOT), pickFrom(RARE_LOOT)];
  }
  if (totalScore > 220) {
    return [pickFrom(RARE_LOOT), pickFrom(COMMON_LOOT)];
  }
  return [pickFrom(COMMON_LOOT), pickFrom(COMMON_LOOT)];
}
